### ISSUE TITLE
BUG: type of <float> * NDArray[int64] is NDArray[int64]

### ISSUE THREAD

**Issue by clo-vis** [numpy/numpy#21427]
### Describe the issue:

The type of <float> * NDArray[int64] is not correctly inferred.

### Reproduce the code example:

```python
from numpy.typing import NDArray
from numpy import int64, float64

def my_fun(array: NDArray[int64]) -> NDArray[float64]:
    return 0.01 * array
```

### Error message:

```shell
mypy complains: 

Incompatible return value type (got "ndarray[Any, dtype[signedinteger[Any]]]", expected "ndarray[Any, dtype[floating[_64Bit]]]")  [return-value]
```

### NumPy/Python version information:

numpy: 1.22.3
python: 3.10.4
mypy: 0.950

**Comment by BvB93** [comment]
There is something wrong with your example, as `NDArray` lives in the `numpy.typing` namespace.
I suspect this might be the cause of your issue, as otherwise (when fixing the import path) I'm unable to reproduce the error.

**Comment by clo-vis** [comment]
I fixed the imports.
What I forgot to tell: I ran mypy using --strict (which enables all optional error checking flags).
Then I narrowed down the flag that is actually responsible for the error: 

$ _mypy --no-implicit-reexport test.py_
test.py:6: error: Incompatible return value type (got "ndarray[Any, dtype[signedinteger[Any]]]", expected "ndarray[Any, dtype[floating[_64Bit]]]")
Found 1 error in 1 file (checked 1 source file)

**Comment by BvB93** [comment]
> Then I narrowed down the flag that is actually responsible for the error: `--no-implicit-reexport`

Ah, so that's what's going on. By the looks of it this is the same problem as https://github.com/numpy/numpy/issues/21214, an unintended side-effect of have the implicit re-exports of private types were managed. 

In any case, the issue is already fixed for the upcoming numpy 1.22.4 release.

**Comment by BvB93** [comment]
I'm going to close this now, as this issue has been already been fixed for the numpy 1.22.4 (and 1.23.0) release.
Feel free to re-open though if the problem somehow persists.
